### PR TITLE
Tracking doc: rewrite `@wdio/firefox-profile-service` package into TypeScript

### PR DESCRIPTION
**Pull request by suniljaiswal01**
## Proposed changes

This is a tracking issue for organising the effort of re-writing the code base from vanilla JS into TypeScript. If you are reading this and are interested getting involved, this is a great chance to do so. Here are the rules:

if you are interested converting source files from /packages/wdio-firefox-profile-service into TypeScript please comment on this thread and mention the files you want to convert to avoid duplicate work
there is no reason to convert the whole package at once - smaller PRs make it easier to review the code
if you make a PR, reference this issue so we can keep track of the progress
if you make a PR, label it with TypeScript and the package name wdio-firefox-profile-service so we can easier find all PRs
if you make a PR, assign it to the TypeScript Rewrite project

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)
#5836 

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/wdio-firefox-profile-service/src/launcher.ts [hunk 1]
```diff
@@ -15,7 +17,8 @@ export default class FirefoxProfileLauncher {
         }
 
         if (this.options.profileDirectory) {
-            this.profile = await promisify(Profile.copy)(this.options.profileDirectory)
+            const profileCopy : (options: WebdriverIO.ServiceOption | string) => Promise<any> = promisify(Profile.copy)
```
Comment (christian-bromann) [inline packages/wdio-firefox-profile-service/src/launcher.ts:20]: I don't think a `Promise<any>` is returned here, it should be a Firefox options profile.

```suggestion
            const profileCopy: (options: WebdriverIO.ServiceOption | string) => Promise<any> = promisify(Profile.copy)
```

File: packages/wdio-firefox-profile-service/src/launcher.ts [hunk 2]
```diff
@@ -51,21 +54,21 @@ export default class FirefoxProfileLauncher {
         this.profile.updatePreferences()
     }
 
-    async _buildExtension(capabilities) {
-        const zippedProfile = await promisify(this.profile.encoded.bind(this.profile))()
+    async _buildExtension(capabilities: WebDriver.Capabilities[] | Record<string, WebDriver.Capabilities & WebDriver.Options>) {
```
Comment (christian-bromann) [inline packages/wdio-firefox-profile-service/src/launcher.ts:57]:
```suggestion
    async private _buildExtension(capabilities: WebDriver.Capabilities[] | Record<string, WebDriver.Capabilities & WebDriver.Options>) {
```

**Comment by christian-bromann** [comment]
@suniljaiswal01 anything I can help to land this?

**Comment by suniljaiswal01** [comment]
> @suniljaiswal01 anything I can help to land this?

@christian-bromann  I am facing issue while giving type to Capabilities paramater inside the Constructor.

https://github.com/suniljaiswal01/webdriverio/blob/typescript-firefox-profile/packages/wdio-firefox-profile-service/src/launcher.ts#L11  > this line
other capabilities types  in this class are depend on the type we are providing in Constructor.
Can you help me in identifying what should be the correct typing over here

**Comment by christian-bromann** [comment]
> Can you help me in identifying what should be the correct typing over here

Please check the other packages. Usually you should be able to access these types by adding `webdriverio` to the `"types"` properties in the subpackage tsconfig:

e.g. in `packages/devtools/tsconfig.json`:

```json
{
    "extends": "../../tsconfig",
    "compilerOptions": {
        "baseUrl": ".",
        "outDir": "./build",
        "rootDir": "./src",
        "types": ["webdriverio", "@wdio/protocols"],
    },
    "include": [
        "src/**/*"
    ]
}
```

Then you can use the types in all files via, e.g. in `packages/devtools/src/index.ts`

```ts
static async newSession (options: WebDriver.Options = {}, modifier?: Function, userPrototype = {}, customCommandWrapper?: Function) {
        const params: WebDriver.Options = validateConfig(DEFAULTS, options)
```

Can you try this?

**Comment by christian-bromann** [comment]
Closing in favor of #5836

**Pull request closed without merging**